File: runtime/python/prompty/prompty/harness/replay_verifier.py
"""Verify normalized replay journals against generated harness contracts."""

from __future__ import annotations

from ..model import (
    ReplayJournalRecord,
    ReplayMismatch,
    ReplayVerificationRequest,
    ReplayVerificationResult,
)


def _comparable(record: ReplayJournalRecord | None) -> dict | None:
    return None if record is None else record.save()


class ReferenceReplayVerifier:
    """Compare expected and actual normalized replay journal records."""

    def verify(self, request: ReplayVerificationRequest) -> ReplayVerificationResult:
        """Return generated replay verification result data."""
        expected = request.expected or []
        actual = request.actual or []
        mismatches: list[ReplayMismatch] = []

        for index in range(max(len(expected), len(actual))):
            expected_record = expected[index] if index < len(expected) else None
            actual_record = actual[index] if index < len(actual) else None
            if _comparable(expected_record) != _comparable(actual_record):
                if expected_record is None:
                    message = "Unexpected extra replay record"
                elif actual_record is None:
                    message = "Missing replay record"
                else:
                    message = "Replay record mismatch"
                mismatches.append(
                    ReplayMismatch(
                        index=index,
                        expected=expected_record,
                        actual=actual_record,
                        message=message,
                    )
                )

        return ReplayVerificationResult(
            status="passed" if not mismatches else "failed",
            mismatches=mismatches,
            expected_count=len(expected),
            actual_count=len(actual),
        )

File: runtime/go/prompty/model/replay_verifier.go
package prompty

import (
	"encoding/json"
	"fmt"
)

// ReferenceReplayVerifier compares normalized replay journal records.
type ReferenceReplayVerifier struct{}

// Verify returns a generated replay verification result.
func (ReferenceReplayVerifier) Verify(request ReplayVerificationRequest) ReplayVerificationResult {
	expected := request.Expected
	actual := request.Actual
	max := len(expected)
	if len(actual) > max {
		max = len(actual)
	}
	mismatches := []ReplayMismatch{}

	for index := 0; index < max; index++ {
		var expectedRecord *ReplayJournalRecord
		var actualRecord *ReplayJournalRecord
		if index < len(expected) {
			expectedRecord = &expected[index]
		}
		if index < len(actual) {
			actualRecord = &actual[index]
		}
		if comparableReplayRecord(expectedRecord) != comparableReplayRecord(actualRecord) {
			message := "Replay record mismatch"
			if expectedRecord == nil {
				message = "Unexpected extra replay record"
			} else if actualRecord == nil {
				message = "Missing replay record"
			}
			mismatches = append(mismatches, ReplayMismatch{
				Index:    int32(index),
				Expected: expectedRecord,
				Actual:   actualRecord,
				Message:  message,
			})
		}
	}

	status := ReplayVerificationStatusPassed
	if len(mismatches) > 0 {
		status = ReplayVerificationStatusFailed
	}
	return ReplayVerificationResult{
		Status:        status,
		Mismatches:    mismatches,
		ExpectedCount: int32(len(expected)),
		ActualCount:   int32(len(actual)),
	}
}

func comparableReplayRecord(record *ReplayJournalRecord) string {
	if record == nil {
		return "<nil>"
	}
	bytes, err := json.Marshal(record.Save(NewSaveContext()))
	if err != nil {
		return fmt.Sprintf("<error:%v>", err)
	}
	return string(bytes)
}
